cps/tasks/download.py

```diff
@@ -39,53 +40,70 @@ def run(self, worker_thread):
             # Execute the download process using process_open
             try:
                 p = process_open(subprocess_args, newlines=True)
-
-                # Define the pattern for the subprocess output
-                pattern_analyze = r"Running ANALYZE"
-                pattern_download = r"'action': 'download'"
-
                 while p.poll() is None:
                     line = p.stdout.readline()
+                    self.message = "Extracting metadata..."
                     if line:
                         log.info(line)
-                        if pattern_analyze in line:
-                            log.info("Matched output (ANALYZE): %s", line)
-                            self.progress = 0.1
-                        if pattern_download in line:
-                            log.info("Matched output (download): %s", line)
-                            self.progress = 0.5
-                            
+                        if "creating" in line:
+                            self.progress += 0.025
+                        log.info("Download progress: %s", self.progress)
+                        if "Running ANALYZE" in line:
+                            self.progress = 0
+                            line = p.stdout.readline()
+                        if "extracting" in line:
+                            self.message = "Downloading media..."
+                            self.progress += 0.077
+                        elif "downloading" in line:
+                            self.progress += 0.077
+                        elif "download" in line:
+                            self.progress += 0.077
+                        elif "merging" in line:
+                            self.progress += 0.077
+                        elif "adding" in line:
+                            self.progress += 0.075
+                            break
+                        else:
+                            self.progress += 0.077
+                        log.info("Download progress: %s", self.progress)
```

Comment:
These lines need more explaining:

- How were the 8 numbers chosen?
- Why is 0.077 used 5 times, and 0.075 used once?
- Why is a `break` statement now needed with "adding", unlike in the past?
- Does 0.075 do anything at all, if it breaks out of the while loop, and is overwritten soon after (on Line 102) `self.progress = 1`
- Can the solitary test for "creating" (on Line 48) possibly be made a part of the if-elif-elif-elif-elif-else structure further below?

Reply:
I refactored it and got rid of pattern matching. The logging will be more aligned with #68. However it builds on a customization made to xklb's underlying yt_dlp code, in which I use progress hooks.

Reply:
> builds on a customization made to xklb's underlying yt_dlp code, in which I use progress hooks

Reply:
> builds on a customization made to xklb's underlying yt_dlp code

What xklb or yt-dlp code are you talking about, to help others understand?

> in which I use progress hooks

What line number in your code?  (Thanks!)

Reply:
I am talking about this changes:
https://github.com/chapmanjacobd/library/compare/main...deldesir:library:patch-1

Reply:
> [chapmanjacobd/library@main...deldesir:library:patch-1](https://github.com/chapmanjacobd/library/compare/main...deldesir:library:patch-1)

@chapmanjacobd just FYI! 🏗️

Reply:
> I am talking about this changes: [chapmanjacobd/library@main...deldesir:library:main](https://github.com/chapmanjacobd/library/compare/main...deldesir:library:main)

@deldesir please clarify:

- Are you asking everyone to change how they [install IIAB with Calibre-Web](https://github.com/iiab/calibre-web/wiki#wrench-installation) ?

- Are you asking them to customize [xklb](https://github.com/iiab/calibre-web/wiki#gear-back-end) code, after [pipx install xklb](https://github.com/iiab/iiab/blob/42d580ef6520b2fba9d5bef78f7977fb50ee49af/roles/calibre-web/tasks/install.yml#L75) ?

- Please spell out precisely (AI²) whatever you are suggesting (Thanks!)

AI² = Articulate Assumptions, Intentions & Intuition

Reply:
No need to customize xklb. IIAB install instructions remain the same.